### PR TITLE
VB generators

### PR DESCRIPTION
**Pull request by chsienki**
Enable generators in the Visual Basic compiler

### CODE REVIEW

File: src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerFileReference.cs
Comment (chsienki) [inline src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerFileReference.cs:130]: 📝@333fred I'd like to run this by the API compat review board.

File: src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerFileReference.cs
Comment (chsienki) [inline src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerFileReference.cs:116]: 📝I'm not sure if it was ever intentional to return duplicate analyzer references, but I've ensured we keep the previous behavior.

File: src/Compilers/Core/Portable/SourceGeneration/GeneratorAttribute.cs
Comment (chsienki) [inline src/Compilers/Core/Portable/SourceGeneration/GeneratorAttribute.cs:14]: ❓ This is the same layout as the `DiagnosticAnalyzer` attribute, so that we can re-use the lookup logic. Should we enforce this via some stronger mechanism like inheritance (does that break binary compat?) or are we confident with tests covering it working as expected?

**Comment by chsienki** [comment]
@dotnet/roslyn-compiler for review (/cc @jasonmalinowski for IDE). 

Hopefully it's fairly clear what's going on, but you may find it easier to review commit-by-commit, as I tried to do the work in small distinct stages to make review easier.

File: src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerReference.cs
Comment (jaredpar) [inline src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerReference.cs:76]:
```suggestion
        public virtual ImmutableArray<ISourceGenerator> GetGenerators(string language) => ImmutableArray<ISourceGenerator>.Empty; 
```

Nit

File: src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerReference.cs
Comment (jaredpar) [inline src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerReference.cs:70]:
```suggestion
        public virtual ImmutableArray<ISourceGenerator> GetGenerators() => ImmutableArray<ISourceGenerator>.Empty;
```

Nit

File: src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerReference.cs
Comment (jaredpar) [inline src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerReference.cs:67]:
```suggestion
        public virtual ImmutableArray<ISourceGenerator> GetGeneratorsForAllLanguages() => ImmutableArray<ISourceGenerator>.Empty;
```

Nit

File: src/Compilers/Core/Portable/SourceGeneration/AdditionalSourcesCollection.cs
Comment (jaredpar) [inline src/Compilers/Core/Portable/SourceGeneration/AdditionalSourcesCollection.cs:25]: When dealing with extensions can never remember if we should or should not include the `.`. So `".cs"` vs. `"cs"`. Consider either enforcing one or the other is passed here or normalizing the input to a specific one.

File: src/Compilers/Core/Portable/SourceGeneration/AdditionalSourcesCollection.cs
Comment (jaredpar) [inline src/Compilers/Core/Portable/SourceGeneration/AdditionalSourcesCollection.cs:104]: Based on this it seems like we should assert in the ctor that the provided extension begins with `"."`

File: src/Compilers/Core/Portable/SourceGeneration/GeneratorAttribute.cs
Comment (jaredpar) [reply src/Compilers/Core/Portable/SourceGeneration/GeneratorAttribute.cs:14]: I'm fine with it being separate. I think inheritance may be over doing it a bit here.

File: src/Compilers/Core/Portable/SourceGeneration/GeneratorAttribute.cs
Comment (jaredpar) [inline src/Compilers/Core/Portable/SourceGeneration/GeneratorAttribute.cs:15]:
```suggestion
        /// The source languages to which this generator applies. See <see cref="LanguageNames"/>.
```

File: src/Compilers/Core/Portable/SourceGeneration/GeneratorAttribute.cs
Comment (jaredpar) [inline src/Compilers/Core/Portable/SourceGeneration/GeneratorAttribute.cs:20]:
```suggestion
        /// Attribute constructor used to specify the attached class is a source generator that provides CSharp sources.
```

Crazy aside: is it required to be a `class` or can it say be a `struct`?

File: src/Compilers/VisualBasic/Portable/CommandLine/VisualBasicCompiler.vb
Comment (jaredpar) [inline src/Compilers/VisualBasic/Portable/CommandLine/VisualBasicCompiler.vb:300]:
```suggestion

        Private Protected Overrides Function RunGenerators(input As Compilation, parseOptions As ParseOptions, generators As ImmutableArray(Of ISourceGenerator), analyzerConfigOptionsProvider As AnalyzerConfigOptionsProvider, additionalTexts As ImmutableArray(Of AdditionalText), diagnostics As DiagnosticBag) As Compilation
```

File: src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerFileReference.cs
Comment (jaredpar) [inline src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerFileReference.cs:278]: Could this be a local function?

File: src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerFileReference.cs
Comment (jaredpar) [inline src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerFileReference.cs:273]: Could this be a local function?

**Comment by chsienki** [comment]
Closing/reopening to fix CI

**Review by jasonmalinowski — CHANGES_REQUESTED** [review]
IDE stuff looks good except it also needs a change in src/Features/Core/Portable/Diagnostics/AnalyzerHelper.cs, where we updated in https://github.com/dotnet/roslyn/pull/48400. There's comments with "we don't support VB" that need to be updated with the right diagnostic IDs.

**Comment by paul1956** [comment]
Is this internal or the same as the C# Source Generator feature or something else? If its public is there any VB specific documentation/examples?

File: src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerFileReference.cs
Comment (chsienki) [reply src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerFileReference.cs:130]: In a perfect world, I'd just remove it. Right now I've obsoleted it, and it returns only CSharp generators, which is essentially what it did before.

File: src/Compilers/Core/CodeAnalysisTest/Analyzers/AnalyzerFileReferenceTests.cs
Comment (chsienki) [reply src/Compilers/Core/CodeAnalysisTest/Analyzers/AnalyzerFileReferenceTests.cs:251]: Ah, not sure how that got removed. Good catch

File: src/Compilers/Core/Portable/SourceGeneration/GeneratorAttribute.cs
Comment (chsienki) [reply src/Compilers/Core/Portable/SourceGeneration/GeneratorAttribute.cs:20]: It *has* to be a class. We explicitly set the allowed location of the attribute to be class (see line 11) so loading via AnalyzerReference it can't ever be a struct. 

As a second defense we check that the generator is also an `ISourceGenerator`, so there's no way you can pass a struct via the generator driver in code either.

**Comment by jaredpar** [comment]
@paul1956 

> Is this internal or the same as the C# Source Generator feature or something else? If its public is there any VB specific documentation/examples?

This is the same as C# source generators, just for VB. There aren't any VB specific docs / examples at this point. The faeture doesn't exist until this is merged so can't really have samples yet ;)

The existing C# samples though should be generally applicable to the VB story though. There is no material difference in the compilation machinery here.

File: src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerFileReference.cs
Comment (chsienki) [reply src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerFileReference.cs]: I refactored out the extensions logic to make it simpler and not need it.

File: src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerFileReference.cs
Comment (chsienki) [reply src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerFileReference.cs:263]: I've left these for now as they've got the comments explaining them and it seemed weird embedding them in the ctor.

**Comment by chsienki** [comment]
@jaredpar 
> The existing C# samples though should be generally applicable to the VB story though. There is no material difference in the compilation machinery here.

Do we want to update the cookbook as part of this PR, or are we ok leaving it until later?

**Comment by jaredpar** [comment]
> Do we want to update the cookbook as part of this PR, or are we ok leaving it until later?

I'm indifferent here. Think we need the cookbook updated before this is RTM but I don't think we need to tie it to this PR.

File: src/Compilers/Core/CodeAnalysisTest/Analyzers/AnalyzerFileReferenceTests.cs
Comment (jaredpar) [inline src/Compilers/Core/CodeAnalysisTest/Analyzers/AnalyzerFileReferenceTests.cs:252]:
```suggestion
        [ConditionalFact(typeof(CoreClrOnly), Reason = "Can't load a framework targeting generator, which these are in desktop")]
```

**Comment by AnthonyDGreen** [comment]
> @paul1956
> 
> > Is this internal or the same as the C# Source Generator feature or something else? If its public is there any VB specific documentation/examples?
> 
> This is the same as C# source generators, just for VB. There aren't any VB specific docs / examples at this point. The faeture doesn't exist until this is merged so can't really have samples yet ;)
> 
> The existing C# samples though should be generally applicable to the VB story though. There is no material difference in the compilation machinery here.

You don't say...

**Comment by chsienki** [comment]
@jasonmalinowski I've updated the AnalyzerHelper to include the VB warning, and added the required warning to the VB code itself.

I have a test in the VB layer, but currently can't run it because we're not running VB command line tests on core. That is proving to be an _interesting_ challenge, so I may just open an issue for it and add the test as part of the move later on.

**Comment by chsienki** [comment]
Ok I've added the test, even though it will always be skipped for now, and opened an issue to track enabling the command-line tests for VB.

**Review by jaredpar — APPROVED** [review]

**Comment by paul1956** [comment]
When does this appear in some Preview I can try?

**Comment by chsienki** [comment]
@paul1956 
> When does this appear in some Preview I can try?

When the PR merges it'll be available in the Roslyn nightly NuGet packages. We don't have a definitive date or target release of visual studio that it'll be available by, but we'll be sure to call it out when there is one.

**Comment by paul1956** [comment]
@chsienki when it is available in NuGet can you post minimal instructions on where to download and simple usage?

**Comment by chsienki** [comment]
@dotnet/roslyn-compiler for a second review please :)

File: src/Compilers/Core/CodeAnalysisTest/Analyzers/AnalyzerFileReferenceTests.cs
Comment (cston) [inline src/Compilers/Core/CodeAnalysisTest/Analyzers/AnalyzerFileReferenceTests.cs:259]: Consider using: `AssertEx.SetEqual(new[] { "TestGenerator", "BaseGenerator", ... }, typeNames);` here and below.

File: src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerFileReference.cs
Comment (cston) [inline src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerFileReference.cs:257]: >addition [](start = 67, length = 8)

"additional"?

File: src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerFileReference.cs
Comment (cston) [inline src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerFileReference.cs:266]: >addition [](start = 67, length = 8)

"additional"?

File: src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerFileReference.cs
Comment (cston) [inline src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerFileReference.cs:327]: >TAttribute [](start = 18, length = 10)

Is this type parameter used outside of the constructor? If not, consider passing the namespace and name explicitly to the constructor, or create a generic `static Extensions<TExtension> Create<TExtension, TAttribute>(...)` factory method.

File: src/Compilers/Core/CodeAnalysisTest/Analyzers/AnalyzerFileReferenceTests.cs
Comment (jasonmalinowski) [inline src/Compilers/Core/CodeAnalysisTest/Analyzers/AnalyzerFileReferenceTests.cs:256]: And if you do follow @cston's advice below, delete this assert; nothing more fun than this accidentally returning more than it should and the failure is "Expected: 10, actual: 11" and you have no idea what happened.

File: src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerFileReference.cs
Comment (cston) [inline src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerFileReference.cs:370]: >builder.Contains [](start = 50, length = 16)

It looks like this is O(n^2) where n is the number of extensions. Consider adding all analyzers as before and check for duplicates outside the loop: `return builder.ToImmutable().Distinct(comparer);`

File: src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerFileReference.cs
Comment (jasonmalinowski) [reply src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerFileReference.cs:112]: Should we just break this? I admit I can't think of anything that would care, and honestly more that might be broken. :-)

**Review by jasonmalinowski — APPROVED** [review]

File: src/Compilers/CSharp/Test/Semantic/SourceGeneration/AdditionalSourcesCollectionTests.cs
Comment (cston) [inline src/Compilers/CSharp/Test/Semantic/SourceGeneration/AdditionalSourcesCollectionTests.cs:57]: >sources[0] [](start = 24, length = 10)

Consider using `sources.Single()` here and in test above.

File: src/Compilers/VisualBasic/Test/CommandLine/CommandLineTests.vb
Comment (cston) [inline src/Compilers/VisualBasic/Test/CommandLine/CommandLineTests.vb:10308]: >generatorSource [](start = 93, length = 15)

It looks like the source is C#.

Comment (chsienki) [reply src/Compilers/VisualBasic/Test/CommandLine/CommandLineTests.vb:10308]: It is, generators themselves can be written in any language. I'll switch it over to VB to make it consistent though.

File: src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerFileReference.cs
Comment (chsienki) [reply src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerFileReference.cs:327]: I refactored it just pass in the type as a param instead.

File: src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerFileReference.cs
Comment (chsienki) [reply src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerFileReference.cs:370]: The irony that I originally did it this way to avoid the duplicate work of calling distinct afterwards. Good catch.

File: src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerFileReference.cs
Comment (chsienki) [reply src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerFileReference.cs:112]: Yeah, I have no idea. It *is* a public API though, and it's been around for at least 6 years according to blame.

File: src/Compilers/Core/CodeAnalysisTest/Analyzers/AnalyzerFileReferenceTests.cs
Comment (cston) [inline src/Compilers/Core/CodeAnalysisTest/Analyzers/AnalyzerFileReferenceTests.cs:274]: `Assert.DoesNotContain()` is no longer necessary.

File: src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerFileReference.cs
Comment (cston) [inline src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerFileReference.cs:383]: >static [](start = 23, length = 6)

`readonly`

File: src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerFileReference.cs
Comment (cston) [inline src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerFileReference.cs:385]: >x is null && y is null || x is object && y is object && x.GetType().Equals(y.GetType() [](start = 68, length = 86)

Can use `object.Equals(x?.GetType(), y?.GetType())`.

**Review by cston — APPROVED** [review]

**Pull request merged**
